clustercontroller-core/src/main/java/com/yahoo/vespa/clustercontroller/core/restapiv2/requests/SetNodeStateRequest.java

```diff
@@ -95,15 +97,92 @@ static SetResponse setWantedState(
-            stateListener.handleNewWantedNodeState(nodeInfo, newWantedState);
+            setNewWantedState(nodeInfo, newWantedState, stateListener);
         }
 
-        // wasModified is true if the new/current State equals the wanted state in the request.
-        boolean wasModified = result.settingWantedStateIsAllowed() || result.wantedStateAlreadySet();
-        // If the state was successfully set, just return an "ok" message back.
-        String reason = wasModified ? "ok" : result.getReason();
-        return new SetResponse(reason, wasModified);
+        // True if the wanted state was or has just been set to newWantedState
+        boolean success = result.settingWantedStateIsAllowed() || result.wantedStateAlreadySet();
+
+        if (success && condition == SetUnitStateRequest.Condition.SAFE && nodeInfo.isStorage()) {
+            // In safe-mode, setting the storage node must be accompanied by changing the state
+            // of the distributor. E.g. setting the storage node to maintenance may cause
+            // feeding issues unless distributor is also set down.
+
+            ensureDistributorWantedState(cluster, nodeInfo.getNodeIndex(), newWantedState, stateListener);
+        }
+
+        return success;
+    }
+
+    /**
+     * Set the wanted state on the distributor to something appropriate given the storage is being
+     * set to (or is equal to) newStorageWantedState.
+     */
+    private static void ensureDistributorWantedState(ContentCluster cluster,
+                                                     int index,
+                                                     NodeState newStorageWantedState,
+                                                     NodeStateOrHostInfoChangeHandler stateListener) {
+        Node distributorNode = new Node(NodeType.DISTRIBUTOR, index);
+        NodeInfo nodeInfo = cluster.getNodeInfo(distributorNode);
+        if (nodeInfo == null) {
+            throw new IllegalStateException("Missing distributor at index " +
+                    distributorNode.getIndex());
+        }
+
+        State newState;
+        switch (newStorageWantedState.getState()) {
+            case MAINTENANCE:
+                newState = State.DOWN;
+                break;
+            case RETIRED:
+                newState = State.UP;
+                break;
+            default:
+                newState = newStorageWantedState.getState();
+                if (!newState.validWantedNodeState(distributorNode.getType())) {
+                    throw new IllegalStateException("Distributor cannot be set to wanted state " +
+                            newState);
+                }
+        }
+
+        NodeState newWantedState = new NodeState(distributorNode.getType(), newState);
+        newWantedState.setDescription(newStorageWantedState.getDescription());
+
+        NodeState currentWantedState = nodeInfo.getUserWantedState();
+        if (newWantedState.getState() != currentWantedState.getState()) {
```

Comment:
This will not update the wanted state if only the description has changed. Should we allow state changes that only change the description stored in ZooKeeper? Might be useful to be able to safe-mode transition a node from "Down: something fishy..." -> "Down: node literally on fire! do not use!!"

Reply:
Yeah, I had that in an earlier revision but removed it since it seemed a bit ugly, since no 'equals' method can be used. Given your concern I have added it back.